File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
from operator import mul
from functools import reduce

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        zeronum = nums.count(0)

        if zeronum > 1:
            return [0] * len(nums)
        elif zeronum == 1:
            acc = reduce(mul, (it for it in nums if it != 0) )
            return [acc if it == 0 else 0 for it in nums]
        else:
            acc = reduce(mul, (it for it in nums if it != 0) )
            return [acc // it for it in nums]

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.ts
function productExceptSelf(nums: number[]): number[] {
	const zerocnt = nums.filter(it => it == 0).length
	if (zerocnt >= 2) {
		return _.range(nums.length).map(_ => 0)
	} else if (zerocnt == 1) {
		const acc = nums.filter(it => it != 0).reduce((acc, it) => acc * it)
		return nums.map(it => it == 0 ? acc : 0)
	} else {
		const acc = nums.reduce((acc, it) => acc * it)
		return nums.map(it => acc / it)
	}
};